File: packages/rank/android/src/main/java/io/capkit/rank/RankConfig.kt
package io.capkit.rank

import android.content.Context
import com.getcapacitor.Plugin

/**
 * Plugin configuration holder for the Rank plugin.
 *
 * This class is responsible for reading and parsing static configuration values
 * defined under the `plugins.Rank` key in `capacitor.config.ts`.
 *
 * Architectural rules:
 * - Read once during plugin initialization in the load() phase.
 * - Configuration values are read-only at runtime.
 * - Consumed only by native code.
 */
class RankConfig(
  plugin: Plugin,
) {
  // ---------------------------------------------------------------------------
  // Properties
  // ---------------------------------------------------------------------------

  /**
   * Android application context.
   * Accessible for native implementation components that require system services.
   */
  val context: Context = plugin.context

  /**
   * Enables verbose native logging via RankLogger.
   *
   * When true, additional debug information and lifecycle events are printed to Logcat.
   * Default: false
   */
  val verboseLogging: Boolean

  /**
   * The Android Package Name used for Play Store redirection.
   *
   * If provided, this value overrides the host application's package name
   * during store navigation.
   * Default: null (falls back to host app package)
   */
  val androidPackageName: String?

  /**
   * Global policy for review request resolution.
   *
   * If true, the `requestReview` method resolves the promise immediately
   * without waiting for the Google Play review flow to complete.
   * Default: false
   */
  val fireAndForget: Boolean

  // ---------------------------------------------------------------------------
  // Initialization
  // ---------------------------------------------------------------------------

  init {
    // Access the plugin-specific configuration object
    val config = plugin.getConfig()

    // Extract verboseLogging flag
    verboseLogging = config.getBoolean("verboseLogging", false)

    // Extract and validate the Android Package Name
    val apm = config.getString("androidPackageName")
    androidPackageName = if (!apm.isNullOrBlank()) apm else null

    // Extract the fireAndForget resolution policy
    fireAndForget = config.getBoolean("fireAndForget", false)
  }
}

File: packages/rank/ios/Sources/RankPlugin/config/RankConfig.swift
import Foundation
import Capacitor

/**
 * Plugin configuration container for the Rank plugin.
 *
 * This struct is responsible for reading and exposing static configuration
 * values defined under the `Rank` key in `capacitor.config.ts`.
 *
 * Architectural rules:
 * - Read once during plugin initialization (`load()`).
 * - Treated as immutable runtime input.
 * - Consumed only by native code (never by JavaScript).
 */
public struct RankConfig {

    // MARK: - Configuration Keys

    /**
     * Internal structure to maintain consistent configuration keys.
     */
    private struct Keys {
        static let verboseLogging = "verboseLogging"
        static let appleAppId = "appleAppId"
        static let fireAndForget = "fireAndForget"
    }

    // MARK: - Public Properties

    /**
     * Enables verbose native logging via RankLogger.
     *
     * When enabled, additional debug information is printed to the Xcode console.
     * Default: false
     */
    public let verboseLogging: Bool

    /**
     * Optional Apple App ID for App Store redirection.
     *
     * If provided, the plugin uses this ID to direct users to the specific review page.
     * Default: nil
     */
    let appleAppId: String?

    /**
     * Global policy for review request resolution.
     *
     * If true, the plugin resolves promises immediately without awaiting OS feedback.
     * Note: On iOS, this is the default behavior as StoreKit provides no completion callback.
     * Default: false
     */
    let fireAndForget: Bool

    // MARK: - Private Defaults

    private static let defaultVerboseLogging: Bool = false
    private static let defaultAppleAppId: String? = nil
    private static let defaultFireAndForget: Bool = false

    // MARK: - Initialization

    /**
     * Initializes the configuration by reading values from the Capacitor bridge.
     *
     * - Parameter plugin: The CAPPlugin instance used to access typed configuration via `getConfig()`.
     */
    init(plugin: CAPPlugin) {
        let config = plugin.getConfig()

        // Extract verboseLogging with fallback to default
        self.verboseLogging = config.getBoolean(
            Keys.verboseLogging,
            Self.defaultVerboseLogging
        )

        // Extract optional appleAppId string
        self.appleAppId = config.getString(
            Keys.appleAppId,
            Self.defaultAppleAppId
        )

        // Extract fireAndForget with fallback to default
        self.fireAndForget = config.getBoolean(
            Keys.fireAndForget,
            Self.defaultFireAndForget
        )
    }
}
